File: leetcode/merge-sorted-array.rs
/* -------------------------------------------------------------------------- */
/*              https://leetcode.com/problems/merge-sorted-array/             */
/* -------------------------------------------------------------------------- */

fn merge(nums1: &mut Vec<i32>, m: i32, nums2: &mut Vec<i32>, n: i32) {
    let mut i = m - 1;
    let mut j = n - 1;
    let mut k = m + n - 1;

    while i >= 0 && j >= 0 {
        if nums1[i as usize] >= nums2[j as usize] {
            nums1[k as usize] = nums1[i as usize];
            i -= 1;
            k -= 1;
        } else {
            nums1[k as usize] = nums2[j as usize];
            j -= 1;
            k -= 1;
        }
    }

    while i >= 0 {
        nums1[k as usize] = nums1[i as usize];
        i -= 1;
        k -= 1;
    }

    while j >= 0 {
        nums1[k as usize] = nums2[j as usize];
        j -= 1;
        k -= 1;
    }
}

fn main() {
    let mut vec1 = vec![1, 2, 6, 0, 0];
    let mut vec2 = vec![3, 4];
    merge(&mut vec1, 3, &mut vec2, 2);
    println!("{:?}", vec1);
}

File: leetcode/cpp/merge-sorted-array.cpp
/* -------------------------------------------------------------------------- */
/*              https://leetcode.com/problems/merge-sorted-array/             */
/* -------------------------------------------------------------------------- */

#include <iostream>
#include <vector>

using namespace std;

class Solution {
   public:
    /**
     * Try to merge them from backs.
     **/
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int i = m - 1, j = n - 1;
        int k = m + n - 1;

        while (i >= 0 && j >= 0) {
            if (nums1[i] >= nums2[j]) {
                nums1[k] = nums1[i];
                k--;
                i--;
            } else {
                nums1[k] = nums2[j];
                k--;
                j--;
            }
        }

        while (i >= 0) {
            nums1[k] = nums1[i];
            k--;
            i--;
        }

        while (j >= 0) {
            nums1[k] = nums2[j];
            k--;
            j--;
        }
    }
};

int main() {
    vector<int> nums1({1, 2, 3, 0, 0, 0});
    vector<int> nums2({2, 5, 6});
    Solution().merge(nums1, 3, nums2, 3);
    for (int i = 0; i < nums1.size(); i++) {
        cout << nums1[i] << " ";
    }
    cout << "\n";
}